File: dd_gazebo_ws/src/drone_nav/src/DroneStateManager.cpp
#include "rclcpp/rclcpp.hpp"
#include "px4_msgs/msg/vehicle_command.hpp"
#include "px4_msgs/msg/vehicle_command_ack.hpp"
#include "px4_msgs/msg/vehicle_status.hpp"
#include "std_msgs/msg/string.hpp"
#include <memory>

using namespace std::chrono_literals;

class DroneStateManager : public rclcpp::Node {
public:
    DroneStateManager()
    : Node("drone_state_manager_node")
    {
        // Publishers and Subscribers
        command_publisher_ = this->create_publisher<px4_msgs::msg::VehicleCommand>(
            "/fmu/in/vehicle_command", 10);

        command_ack_subscription_ = this->create_subscription<px4_msgs::msg::VehicleCommandAck>(
            "/fmu/out/vehicle_command_ack", 10,
            std::bind(&DroneStateManager::vehicleCommandAckCallback, this, std::placeholders::_1));

        vehicle_status_subscription_ = this->create_subscription<px4_msgs::msg::VehicleStatus>(
            "/fmu/out/vehicle_status", 10,
            std::bind(&DroneStateManager::vehicleStatusCallback, this, std::placeholders::_1));

        signal_subscription_ = this->create_subscription<std_msgs::msg::String>(
            "signal", 10,
            std::bind(&DroneStateManager::signalCallback, this, std::placeholders::_1));

        // Initialize state variables
        current_state_ = State::Idle;
        is_armed_ = false;
        awaiting_ack_command_ = 0;

        RCLCPP_INFO(this->get_logger(), "DroneStateManager activated, waiting for signals 'A' or 'B'");

        // Initialize timer for checking disarming state
        timer_ = this->create_wall_timer(
            500ms, std::bind(&DroneStateManager::timerCallback, this));
    }

private:
    enum class State {
        Idle,
        Arming,
        SettingMode,
        TakingOff,
        Hovering,
        Landing,
        WaitUntilDisarmed
    };

    // MAVLink command and result constants
    static constexpr uint16_t MAV_CMD_COMPONENT_ARM_DISARM = 400;
    static constexpr uint16_t MAV_CMD_DO_SET_MODE = 176;
    static constexpr uint16_t MAV_CMD_NAV_TAKEOFF = 22;
    static constexpr uint16_t MAV_CMD_NAV_LAND = 21;
    static constexpr uint8_t VEHICLE_RESULT_ACCEPTED = 0;
    static constexpr uint8_t ARMING_STATE_ARMED = 2;

    void signalCallback(const std_msgs::msg::String::SharedPtr msg)
    {
        if (msg->data == "A") {
            RCLCPP_INFO(this->get_logger(), "Signal 'A' received, initiating takeoff sequence");
            runState(State::Arming);
        } else if (msg->data == "B") {
            RCLCPP_INFO(this->get_logger(), "Signal 'B' received, initiating landing");
            runState(State::Landing);
        } else {
            RCLCPP_WARN(this->get_logger(), "Unknown signal received: '%s'", msg->data.c_str());
        }
    }

    void runState(State state)
    {
        switch (state) {
            case State::Arming:
                arm();
                break;
            case State::SettingMode:
                setOffboardMode();
                break;
            case State::TakingOff:
                sendTakeoffCommand();
                break;
            case State::Landing:
                land();
                break;
            case State::WaitUntilDisarmed:
                waitUntilDisarmed();
                break;
            default:
                RCLCPP_WARN(this->get_logger(), "Unknown state");
                break;
        }
    }

    void arm()
    {
        if (current_state_ != State::Idle) {
            RCLCPP_WARN(this->get_logger(), "Cannot arm: Drone is not in Idle state");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Sending arm command");
        auto cmd = px4_msgs::msg::VehicleCommand();
        cmd.command = MAV_CMD_COMPONENT_ARM_DISARM;
        cmd.param1 = 1.0;  // Arm
        fillVehicleCommand(cmd);
        command_publisher_->publish(cmd);
        current_state_ = State::Arming;
        awaiting_ack_command_ = MAV_CMD_COMPONENT_ARM_DISARM;
    }

    void setOffboardMode()
    {
        if (current_state_ != State::Arming) {
            RCLCPP_WARN(this->get_logger(), "Cannot set mode: Drone is not in Arming state");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Setting offboard mode");
        auto cmd = px4_msgs::msg::VehicleCommand();
        cmd.command = MAV_CMD_DO_SET_MODE;
        cmd.param1 = 1.0;  // Base mode
        cmd.param2 = 6.0;  // Custom mode: PX4_CUSTOM_MAIN_MODE_OFFBOARD
        fillVehicleCommand(cmd);
        command_publisher_->publish(cmd);
        current_state_ = State::SettingMode;
        awaiting_ack_command_ = MAV_CMD_DO_SET_MODE;
    }

    void sendTakeoffCommand()
    {
        if (current_state_ != State::SettingMode) {
            RCLCPP_WARN(this->get_logger(), "Cannot take off: Drone is not in SettingMode state");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Sending takeoff command");
        auto cmd = px4_msgs::msg::VehicleCommand();
        cmd.command = MAV_CMD_NAV_TAKEOFF;
        cmd.param7 = 10.0;  // Desired altitude in meters
        fillVehicleCommand(cmd);
        command_publisher_->publish(cmd);
        current_state_ = State::TakingOff;
        awaiting_ack_command_ = MAV_CMD_NAV_TAKEOFF;
    }

    void land()
    {
        if (current_state_ != State::Hovering && current_state_ != State::TakingOff) {
            RCLCPP_WARN(this->get_logger(), "Cannot land: Drone is not in Hovering or TakingOff state");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Sending land command");
        auto cmd = px4_msgs::msg::VehicleCommand();
        cmd.command = MAV_CMD_NAV_LAND;
        fillVehicleCommand(cmd);
        command_publisher_->publish(cmd);
        current_state_ = State::Landing;
        awaiting_ack_command_ = MAV_CMD_NAV_LAND;
    }

    void waitUntilDisarmed()
    {
        if (current_state_ != State::Landing) {
            RCLCPP_WARN(this->get_logger(), "Cannot wait until disarmed: Drone is not in Landing state");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Waiting until drone is disarmed");
        current_state_ = State::WaitUntilDisarmed;
    }

    void vehicleCommandAckCallback(const px4_msgs::msg::VehicleCommandAck::SharedPtr msg)
    {
        if (msg->command != awaiting_ack_command_) {
            // Ignore acknowledgments for commands we're not tracking
            return;
        }

        if (msg->result == VEHICLE_RESULT_ACCEPTED) {
            if (msg->command == MAV_CMD_COMPONENT_ARM_DISARM) {
                RCLCPP_INFO(this->get_logger(), "Arm command accepted");
                runState(State::SettingMode);
            } else if (msg->command == MAV_CMD_DO_SET_MODE) {
                RCLCPP_INFO(this->get_logger(), "Set mode command accepted");
                runState(State::TakingOff);
            } else if (msg->command == MAV_CMD_NAV_TAKEOFF) {
                RCLCPP_INFO(this->get_logger(), "Takeoff command accepted");
                current_state_ = State::Hovering;
            } else if (msg->command == MAV_CMD_NAV_LAND) {
                RCLCPP_INFO(this->get_logger(), "Landing command accepted");
                runState(State::WaitUntilDisarmed);
            } else {
                RCLCPP_WARN(this->get_logger(), "Received acknowledgment for unknown command: %u", msg->command);
            }
        } else {
            RCLCPP_ERROR(this->get_logger(), "Command %u rejected with result %u", msg->command, msg->result);
            current_state_ = State::Idle;
        }

        // Reset awaiting acknowledgment command
        awaiting_ack_command_ = 0;
    }

    void vehicleStatusCallback(const px4_msgs::msg::VehicleStatus::SharedPtr msg)
    {
        // Update arming state
        is_armed_ = (msg->arming_state == ARMING_STATE_ARMED);
        if (!is_armed_ && current_state_ == State::WaitUntilDisarmed) {
            RCLCPP_INFO(this->get_logger(), "Drone disarmed, operation complete");
            current_state_ = State::Idle;
        }
    }

    void timerCallback()
    {
        // This timer can be used to perform periodic checks or actions if necessary
        // For now, it's just a placeholder
    }

    void fillVehicleCommand(px4_msgs::msg::VehicleCommand &cmd)
    {
        cmd.target_system = 1;
        cmd.target_component = 1;
        cmd.source_system = 1;
        cmd.source_component = 1;
        cmd.from_external = true;
        cmd.timestamp = this->get_clock()->now().nanoseconds() / 1000;  // Timestamp in microseconds
    }

    // Class members
    rclcpp::Publisher<px4_msgs::msg::VehicleCommand>::SharedPtr command_publisher_;
    rclcpp::Subscription<px4_msgs::msg::VehicleCommandAck>::SharedPtr command_ack_subscription_;
    rclcpp::Subscription<px4_msgs::msg::VehicleStatus>::SharedPtr vehicle_status_subscription_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr signal_subscription_;
    rclcpp::TimerBase::SharedPtr timer_;

    State current_state_;
    bool is_armed_;
    uint16_t awaiting_ack_command_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto drone_state_manager = std::make_shared<DroneStateManager>();
    rclcpp::spin(drone_state_manager);
    rclcpp::shutdown();
    return 0;
}

File: dd_gazebo_ws/src/drone_nav/src/drone_state_manager.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from px4_msgs.msg import VehicleCommand, VehicleCommandAck, VehicleStatus
from enum import Enum


class DroneStateManager(Node):
    class State(Enum):
        Idle = 0
        Arming = 1
        SettingMode = 2
        TakingOff = 3
        Hovering = 4
        Landing = 5
        WaitUntilDisarmed = 6

    def __init__(self):
        super().__init__('drone_state_manager_node')

        # Subscribe to the "signal" topic
        self.signal_subscription = self.create_subscription(
            String,
            'signal',
            self.signal_callback,
            10
        )

        self.get_logger().info("DroneStateManager activated, waiting for signals 'A' or 'B'")

        # Publishers and Subscribers
        self.vehicle_command_publisher = self.create_publisher(
            VehicleCommand,
            '/fmu/in/vehicle_command',
            10
        )

        self.vehicle_command_ack_subscription = self.create_subscription(
            VehicleCommandAck,
            '/fmu/out/vehicle_command_ack',
            self.vehicle_command_ack_callback,
            10
        )

        self.vehicle_status_subscription = self.create_subscription(
            VehicleStatus,
            '/fmu/out/vehicle_status',
            self.vehicle_status_callback,
            10
        )

        # Initialize state variables
        self.current_state = self.State.Idle
        self.is_armed = False
        self.awaiting_ack_command = None  # Tracks which command we're waiting for acknowledgment

        # Define MAVLink command constants
        self.MAV_CMD_COMPONENT_ARM_DISARM = 400
        self.MAV_CMD_DO_SET_MODE = 176
        self.MAV_CMD_NAV_TAKEOFF = 22
        self.MAV_CMD_NAV_LAND = 21

        # MAVLink result constants
        self.VEHICLE_RESULT_ACCEPTED = 0

        # Arming state constant
        self.ARMING_STATE_ARMED = 2

    def signal_callback(self, msg):
        if msg.data == "A":
            self.get_logger().info("Signal 'A' received, initiating takeoff sequence")
            self.run_state(self.State.Arming)
        elif msg.data == "B":
            self.get_logger().info("Signal 'B' received, initiating landing")
            self.run_state(self.State.Landing)
        else:
            self.get_logger().warn(f"Unknown signal received: '{msg.data}'")

    def run_state(self, state):
        if state == self.State.Arming:
            self.arm()
        elif state == self.State.SettingMode:
            self.set_offboard_mode()
        elif state == self.State.TakingOff:
            self.send_takeoff_command()
        elif state == self.State.Landing:
            self.land()
        elif state == self.State.WaitUntilDisarmed:
            self.wait_until_disarmed()
        else:
            self.get_logger().warn("Unknown state")

    def arm(self):
        if self.current_state != self.State.Idle:
            self.get_logger().warn("Cannot arm: Drone is not in Idle state")
            return

        self.get_logger().info("Sending arm command")
        cmd = VehicleCommand()
        cmd.command = self.MAV_CMD_COMPONENT_ARM_DISARM
        cmd.param1 = 1.0  # Arm
        cmd.target_system = 1
        cmd.target_component = 1
        cmd.source_system = 1
        cmd.source_component = 1
        cmd.from_external = True
        cmd.timestamp = self.get_clock().now().nanoseconds // 1000
        self.vehicle_command_publisher.publish(cmd)
        self.current_state = self.State.Arming
        self.awaiting_ack_command = self.MAV_CMD_COMPONENT_ARM_DISARM

    def set_offboard_mode(self):
        if self.current_state != self.State.Arming:
            self.get_logger().warn("Cannot set mode: Drone is not in Arming state")
            return

        self.get_logger().info("Setting offboard mode")
        cmd = VehicleCommand()
        cmd.command = self.MAV_CMD_DO_SET_MODE
        cmd.param1 = 1.0  # Base mode
        cmd.param2 = 6.0  # Custom mode: PX4_CUSTOM_MAIN_MODE_OFFBOARD
        cmd.target_system = 1
        cmd.target_component = 1
        cmd.source_system = 1
        cmd.source_component = 1
        cmd.from_external = True
        cmd.timestamp = self.get_clock().now().nanoseconds // 1000
        self.vehicle_command_publisher.publish(cmd)
        self.current_state = self.State.SettingMode
        self.awaiting_ack_command = self.MAV_CMD_DO_SET_MODE

    def send_takeoff_command(self):
        if self.current_state != self.State.SettingMode:
            self.get_logger().warn("Cannot take off: Drone is not in SettingMode state")
            return

        self.get_logger().info("Sending takeoff command")
        cmd = VehicleCommand()
        cmd.command = self.MAV_CMD_NAV_TAKEOFF
        cmd.param7 = 10.0  # Desired altitude in meters
        cmd.target_system = 1
        cmd.target_component = 1
        cmd.source_system = 1
        cmd.source_component = 1
        cmd.from_external = True
        cmd.timestamp = self.get_clock().now().nanoseconds // 1000  # Microseconds
        self.vehicle_command_publisher.publish(cmd)
        self.current_state = self.State.TakingOff
        self.awaiting_ack_command = self.MAV_CMD_NAV_TAKEOFF

    def land(self):
        if self.current_state != self.State.Hovering and self.current_state != self.State.TakingOff:
            self.get_logger().warn("Cannot land: Drone is not in Hovering or TakingOff state")
            return

        self.get_logger().info("Sending land command")
        cmd = VehicleCommand()
        cmd.command = self.MAV_CMD_NAV_LAND
        cmd.target_system = 1
        cmd.target_component = 1
        cmd.source_system = 1
        cmd.source_component = 1
        cmd.from_external = True
        cmd.timestamp = self.get_clock().now().nanoseconds // 1000  # Microseconds
        self.vehicle_command_publisher.publish(cmd)
        self.current_state = self.State.Landing
        self.awaiting_ack_command = self.MAV_CMD_NAV_LAND

    def wait_until_disarmed(self):
        if self.current_state != self.State.Landing:
            self.get_logger().warn("Cannot wait until disarmed: Drone is not in Landing state")
            return

        self.get_logger().info("Waiting until drone is disarmed")
        self.current_state = self.State.WaitUntilDisarmed

    def vehicle_command_ack_callback(self, msg):
        if msg.command != self.awaiting_ack_command:
            # Ignore acknowledgments for commands we're not tracking
            return

        if msg.result == self.VEHICLE_RESULT_ACCEPTED:
            if msg.command == self.MAV_CMD_COMPONENT_ARM_DISARM:
                self.get_logger().info("Arm command accepted")
                self.run_state(self.State.SettingMode)
            elif msg.command == self.MAV_CMD_DO_SET_MODE:
                self.get_logger().info("Set mode command accepted")
                self.run_state(self.State.TakingOff)
            elif msg.command == self.MAV_CMD_NAV_TAKEOFF:
                self.get_logger().info("Takeoff command accepted")
                self.current_state = self.State.Hovering
            elif msg.command == self.MAV_CMD_NAV_LAND:
                self.get_logger().info("Landing command accepted")
                self.run_state(self.State.WaitUntilDisarmed)
            else:
                self.get_logger().warn(f"Received acknowledgment for unknown command: {msg.command}")
        else:
            self.get_logger().error(f"Command {msg.command} rejected with result {msg.result}")
            self.current_state = self.State.Idle

        # Reset awaiting acknowledgment command
        self.awaiting_ack_command = None

    def vehicle_status_callback(self, msg):
        # Update arming state
        self.is_armed = (msg.arming_state == self.ARMING_STATE_ARMED)
        if not self.is_armed and self.current_state == self.State.WaitUntilDisarmed:
            self.get_logger().info("Drone disarmed, operation complete")
            self.current_state = self.State.Idle

    def on_activate(self):
        self.get_logger().info("DroneStateManager node activated")

    def on_deactivate(self):
        self.get_logger().info("DroneStateManager node deactivated")


def main(args=None):
    rclpy.init(args=args)
    drone_state_manager = DroneStateManager()
    try:
        rclpy.spin(drone_state_manager)
    except KeyboardInterrupt:
        pass
    finally:
        drone_state_manager.destroy_node()
        rclpy.shutdown()
